connector/src/main/scala/com/basho/riak/spark/rdd/connector/SessionCache.scala

```diff
@@ -0,0 +1,204 @@
+            */
+          .addRef()
+      }
+    }
+    CacheBuilder.newBuilder()
+      .maximumSize(cacheSize)
+      .removalListener(new RemovalListener[RiakConnectorConf, CachedSession] {
+        override def onRemoval(notification: RemovalNotification[RiakConnectorConf, CachedSession]): Unit = {
+          val (conf, session) = notification.getKey -> notification.getValue
+
+          logDebug(s"Riak CachedSession was evicted: $conf")
+          session.close()
+        }
+      })
+      .build(cacheLoader)
+  }
+
+  def get(conf: RiakConnectorConf): RiakSession = {
+    wrap(cachedSessions.get(conf).addRef())
+  }
+
+  private def wrap(session: CachedSession): RiakSession = {
+    Proxy.newProxyInstance(
+      session.getClass.getClassLoader,
+      Array(classOf[RiakSession]),
+      new SessionProxy(session)
+    ).asInstanceOf[RiakSession]
+  }
+
+  def shutdown(): Unit = {
+    invalidateAll()
+  }
+
+  def invalidateAll(): Unit = {
+    cachedSessions.invalidateAll()
+  }
+
+  private def createClient(conf: RiakConnectorConf): RiakClient = {
+    lazy val addresses = conf.hosts.map(_.toString)
+    lazy val endpointsStr = addresses.mkString("", ", ", "")
+
+    try {
+      logDebug(s"Attempting to create RiakClient(java) for $addresses")
+      val builder = new RiakNode.Builder()
+        .withMinConnections(conf.minConnections)
+        .withMaxConnections(conf.maxConnections)
+
+
+      val nodes = conf.hosts.map { (h: HostAndPort) =>
+        builder.withRemoteAddress(h.getHost)
+        builder.withRemotePort(h.getPort)
+        builder.build()
+      }
+
+      val ns = nodes.toBuffer
+      val cluster = RiakCluster.builder(ns.asJava).build()
+      cluster.start()
+
+      new RiakClient(cluster)
+    } catch {
+      case e: Throwable =>
+        throw new IOException(
+          s"Failed to create RiakClient(java) for $endpointsStr", e)
+    }
+  }
+}
+
+/**
+  * @author Sergey Galkin <srggal at gmail dot com>
+  * @since 1.2.0
+  */
+private class SessionProxy(session: CachedSession) extends InvocationHandler {
+  private var closed: Boolean = false
+
+  override def invoke(proxy: scala.Any, method: Method, args: Array[AnyRef]): AnyRef = {
+    (method.getName, method.getParameterTypes) match {
+      case ("isClosed", Array()) =>
+        closed.asInstanceOf[AnyRef]
+      case _ =>
+        try {
+          require(!closed, "RiakSession should be active")
+          if ("close".equals(method.getName)) {
+            closed = true
+          }
+          method.invoke(session, args: _*)
+        }
+        catch {
+          case e: InvocationTargetException =>
+            throw e.getCause
+        }
+    }
+  }
+}
+
+/**
+  * @author Sergey Galkin <srggal at gmail dot com>
+  * @since 1.2.0
+  */
+class CachedSession(conf: RiakConnectorConf, riakClient: RiakClient) extends RiakSession with Logging {
+  private var closed: Boolean = false
+  private val refCount: AtomicLong = new AtomicLong(0)
```

Comment:
Don't quite understand what does refCount count? It can be 1 or 0, so simply boolean?